src/libostree/ostree-repo-libarchive.c

```diff
@@ -45,269 +47,685 @@ propagate_libarchive_error (GError      **error,
+aic_import_file (OstreeRepoArchiveImportContext *ctx,
+                 OstreeMutableTree  *parent,
+                 const char         *path,
+                 GFileInfo          *fi,
+                 GCancellable       *cancellable,
+                 GError            **error)
+{
+  const char *name = glnx_basename (path);
+  g_autoptr(GVariant) xattrs = NULL;
+  g_autofree char *csum = NULL;
+
+  if (!aic_get_xattrs (ctx, path, fi, &xattrs, cancellable, error))
+    return FALSE;
+
+  if (!aic_write_file (ctx, fi, xattrs, &csum, cancellable, error))
+    return FALSE;
+
+  if (!ostree_mutable_tree_replace_file (parent, name, csum, error))
+    return FALSE;
+
+  return TRUE;
+}
+
+static void
+aic_add_deferred_hardlink (OstreeRepoArchiveImportContext *ctx,
+                           const char        *hardlink,
+                           DeferredHardlink  *dh)
+{
+  gboolean new_slist;
+  GSList *slist;
+
+  slist = g_hash_table_lookup (ctx->deferred_hardlinks, hardlink);
+  new_slist = slist == NULL;
+
+  slist = g_slist_append (slist, dh);
+
+  if (new_slist)
+    g_hash_table_insert (ctx->deferred_hardlinks, g_strdup (hardlink), slist);
+}
+
+static void
+aic_defer_hardlink (OstreeRepoArchiveImportContext *ctx,
+                    OstreeMutableTree  *parent,
+                    const char         *path,
+                    guint64             size,
+                    const char         *hardlink)
+{
+  DeferredHardlink *dh = g_slice_new (DeferredHardlink);
+  dh->parent = g_object_ref (parent);
+  dh->path = g_strdup (path);
+  dh->size = size;
+
+  aic_add_deferred_hardlink (ctx, hardlink, dh);
+}
+
+static gboolean
+aic_handle_file (OstreeRepoArchiveImportContext *ctx,
+                 OstreeMutableTree  *parent,
+                 const char         *path,
+                 GFileInfo          *fi,
+                 GCancellable       *cancellable,
+                 GError            **error)
+{
+  /* The wonderful world of hardlinks and archives. We have to be very careful
+   * here. Do not assume that if a file is a hardlink, it will have size 0 (e.g.
+   * cpio). Do not assume that if a file will have hardlinks to it, it will have
+   * size > 0. Also do not assume that its nlink param is present (tar) or even
+   * accurate (cpio). Also do not assume that hardlinks follow each other in
+   * order of entries.
+   *
+   * These archives were made to be extracted onto a filesystem, not directly
+   * hashed into an object store. So to be careful, we defer all hardlink
+   * imports until the very end. Nonzero files have to be imported, hardlink or
+   * not, since we can't easily seek back to this position later on.
+   * */
+
+  g_autofree char *hardlink = aic_get_final_entry_hardlink (ctx);
+  guint64 size = g_file_info_get_attribute_uint64 (fi, "standard::size");
+
+  if (hardlink == NULL || size > 0)
+    if (!aic_import_file (ctx, parent, path, fi, cancellable, error))
+      return FALSE;
+
   if (hardlink)
```

Comment:
We need to also add to the list hardlinks which carry a payload so that they can be found later. E.g. in `aic_import_from_hardlink` we correct the initial non-hardlink file to point to the hardlink csum.